bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -587,26 +591,71 @@ void PreProcessor::onMessage<PreProcessReplyMsg>(PreProcessReplyMsg *msg) {
 
 template <typename T>
 void PreProcessor::messageHandler(MessageBase *msg) {
+  if (!msgs_.write_available()) {
+    LOG_ERROR(logger(), "PreProcessor queue is full, returning message");
+    incomingMsgsStorage_->pushExternalMsg(std::unique_ptr<MessageBase>(msg));
+    return;
+  }
   T *trueTypeObj = new T(msg);
   delete msg;
-  if (validateMessage(trueTypeObj)) {
-    onMessage<T>(trueTypeObj);
-  } else {
-    preProcessorMetrics_.preProcReqInvalid.Get().Inc();
-    delete trueTypeObj;
-  }
+  msgs_.push(trueTypeObj);
+  msgLoopSignal_.notify_one();
 }
 
 template <>
 void PreProcessor::messageHandler<PreProcessReplyMsg>(MessageBase *msg) {
+  if (!msgs_.write_available()) {
+    LOG_ERROR(logger(), "PreProcessor queue is full, returning message");
+    incomingMsgsStorage_->pushExternalMsg(std::unique_ptr<MessageBase>(msg));
+    return;
+  }
   PreProcessReplyMsg *trueTypeObj = new PreProcessReplyMsg(msg);
   trueTypeObj->setPreProcessorHistograms(&histograms_);
   delete msg;
-  if (validateMessage(trueTypeObj)) {
-    onMessage(trueTypeObj);
-  } else {
-    preProcessorMetrics_.preProcReqInvalid.Get().Inc();
-    delete trueTypeObj;
+  msgs_.push(trueTypeObj);
+  msgLoopSignal_.notify_one();
+}
+
+void PreProcessor::msgProcessingLoop() {
+  while (!msgLoopDone_) {
+    int count = 0;
+    {
+      std::unique_lock<std::mutex> l(msgLock_);
+      while (!msgLoopDone_ && msgs_.empty()) msgLoopSignal_.wait(l);
```

Comment:
I believe there's a race condition here that could result in indefinite waiting. It's due to not holding the lock when `msgs_.push` is called. I only recognized this because I've made a similar mistake before.

The problem is that `msgs_.empty()` can be checked, then the push thread runs, and signal is triggered. Then the wait is called, which misses the signal. 

This isn't likely a huge problem here, as the next push will trigger the signal, but the code is still technically not correct. This is the problem with using a lock-free queue with a condition-variable.

There are 2 ways to fix this:
 1. Just use a regular queue and hold the lock.
 2. Add a timeout to the wait call and recheck the condition.

Reply:
It's not the `empty` that is the problem, although as you pointed out it's not thread safe and should be replace.. The problem is the point in time between the while condition and the `wait` call. There's no way to guarantee that the signal isn't called from another thread in between those 2 points, resulting in missing the signal. 

With a lock around the push method, it wouldn't be possible to miss the signal because you would have to hold the lock to check empty. If you got the lock and it was empty, then you would wait, but the push could only happen after the wait when the lock was released.

But since you likely don't want an extra lock around the pusher, since you are purposefully using a lock free queue, you can just timeout and re-check periodically in case you miss a signal. In practice this will not happen frequently, as throughput will be high.

Reply:
I dont care about the lock around the pusher, the lock from the consumer is very quick, just to take the number of elements, and taking non locked mutex by the producer is not too bad. But in the sake of safety, I prefer to switch to regular queue. Will post an update to this branch.